pkg/server/bundle/pubmanager/pubmanager.go

```diff
@@ -0,0 +1,202 @@
+// stores through the configured BundlePublisher plugins.
+package pubmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/server/catalog"
+	"github.com/spiffe/spire/pkg/server/datastore"
+	"github.com/spiffe/spire/pkg/server/plugin/bundlepublisher"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/zeebo/errs"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// refreshInterval is the interval to check for an updated trust bundle.
+	refreshInterval = 30 * time.Second
+)
+
+// ManagerConfig is the config for the bundle publishing manager.
+type ManagerConfig struct {
+	BundleLoadedCh chan struct{}
+	Catalog        catalog.Catalog
+	Clock          clock.Clock
+	DataStore      datastore.DataStore
+	Log            logrus.FieldLogger
+	TrustDomain    spiffeid.TrustDomain
+}
+
+// Manager is the manager for bundle publishing.
+type Manager struct {
+	bundleLoadedCh chan struct{}
+	catalog        catalog.Catalog
+	clock          clock.Clock
+	dataStore      datastore.DataStore
+	log            logrus.FieldLogger
+	trustDomain    spiffeid.TrustDomain
+
+	hooks struct {
+		// Test hook used to indicate an attempt to publish a bundle using a
+		// specific bundle publisher.
+		publishResultCh chan *publishResult
+
+		// Test hook used to indicate when the action of publishing a bundle
+		// has finished.
+		publishmentFinishedCh chan error
+	}
+
+	bundle    *common.Bundle
+	bundleMtx sync.RWMutex
+}
+
+// Run runs the bundle publishing manager.
+func (m *Manager) Run(ctx context.Context) error {
+	ticker := m.clock.Ticker(refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.bundleLoadedCh:
+			m.bundleLoadedOrClockTicked(ctx)
+		case <-ticker.C:
+			m.bundleLoadedOrClockTicked(ctx)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (m *Manager) bundleLoadedOrClockTicked(ctx context.Context) {
+	// Log an error on failure unless we're shutting down.
+	if err := m.publishIfChanged(ctx); err != nil && ctx.Err() == nil {
+		m.log.WithError(err).Error("Failed to publish bundle")
+	}
+}
+
+// publishIfChanged checks if the trust bundle has changed in the datastore.
+// If it changed, iterates through the configured bundle publishers and calls
+// PublishBundle with the fetched bundle.
+func (m *Manager) publishIfChanged(ctx context.Context) (err error) {
+	bundlePubllishers := m.catalog.GetBundlePublishers()
+	if len(bundlePubllishers) == 0 {
+		return m.finishPublishment(nil)
+	}
+
+	bundle, err := m.dataStore.FetchBundle(ctx, m.trustDomain.IDString())
+	if err != nil {
+		return m.finishPublishment(fmt.Errorf("failed to fetch bundle from datastore: %w", err))
+	}
+
+	// Check if the bundle has changed.
+	if proto.Equal(bundle, m.getBundle()) {
+		return m.finishPublishment(nil)
+	}
+
+	// Bundle changed, update the current bundle in the manager.
+	m.setBundle(bundle)
```

Comment:
I've been thinking about how errors would be handled in case that publishing to a bundle publisher fails and I think that each publisher probably knows better how to deal with that (retries, etc.). So I think that even if there were errors, the proper action here is to not retry with the same bundle and let the publisher handle how to deal with a failure of a Publish action.

Reply:
One way to automatically handle retries is to just push the bundle into each publisher on every tick interval (we still need the ticker... see my other comment) without comparing the bundle to see if anything has changed. It would cause a little extra work but each publisher could hold onto the last version it published and detect if there were any changes.